lang/funcs/encoding.go

```diff
@@ -0,0 +1,135 @@
+package funcs
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/base64"
+	"fmt"
+	"net/url"
+
+	"github.com/zclconf/go-cty/cty"
+	"github.com/zclconf/go-cty/cty/function"
+)
+
+// Base64DecodeFunc constructs a function that decodes a string containing a base64 sequence.
+var Base64DecodeFunc = function.New(&function.Spec{
+	Params: []function.Parameter{
+		{
+			Name: "str",
+			Type: cty.String,
+		},
+	},
+	Type: function.StaticReturnType(cty.String),
+	Impl: func(args []cty.Value, retType cty.Type) (cty.Value, error) {
+		s := args[0].AsString()
+		sDec, err := base64.StdEncoding.DecodeString(s)
+		if err != nil {
+			return cty.UnknownVal(cty.String), fmt.Errorf("failed to decode base64 data '%s'", s)
+		}
+
+		return cty.StringVal(string(sDec)), nil
```

Comment:
Before returning here, we should validate that `sDec` is valid UTF-8 so that we don't get silent corruption of a binary string later. There's an example of this in the implementation of the `file` function, which has a similar constraint when it's not returning base64.

(This is a new requirement compared to the old implementation, where it would've instead just silently corrupted the result.)

Reply:
Ok, thanks! Is there an alternate function to point users towards, or is this just a plain ol' error (I didn't see anything relevant)?

Reply:
Base64 strings are Terraform's primary representation of raw binary data, so in this case there isn't any alternative to suggest: the data should just _stay_ as base64 and be passed like that to resources. There's already some text in the docs that is intended to explain the situation:

> While we do not recommend manipulating large, raw binary data in the Terraform language, Base64 encoding is the standard way to represent arbitrary byte sequences, and so resource types that accept or return binary data will use Base64 themselves, which avoids the need to encode or decode it directly in most cases.

This is referring indirectly to resource arguments like `user_data_base64` on `aws_instance`, where the resource code itself does the decoding, avoiding that raw binary data passing through the Terraform language code and getting corrupted due to UTF-8 interpretation. When I wrote that I considered linking to a few examples, but in this particular case I decided against it because there are quite a few of these and it will grow over time. Instead, I expect that the provider docs would give pointers on how to work with base64 data in Terraform, to come at the problem from the other direction.